src/drivers/virtio/virtqueue/packed.rs

```diff
@@ -897,12 +897,12 @@ impl DevNotif {
 	/// Reads notification bit (i.e. LSB) and returns value.
 	/// If notifications are enabled returns true, else false.
 	fn is_notif(&self) -> bool {
-		self.raw.flags & (1 << 0) == 0
+		self.raw.flags & 0b11 == 0
```

Comment:
I think at least one of the uses of this method is written with the assumption that it checks only for general notifications: 
https://github.com/hermit-os/kernel/blob/737b1388d9d7a62c01a61499c71564d4f70abb5d/src/drivers/virtio/virtqueue/packed.rs#L973

If I understand correctly, for that one and the remaining two uses 
https://github.com/hermit-os/kernel/blob/737b1388d9d7a62c01a61499c71564d4f70abb5d/src/drivers/virtio/virtqueue/packed.rs#L1001 
https://github.com/hermit-os/kernel/blob/737b1388d9d7a62c01a61499c71564d4f70abb5d/src/drivers/virtio/virtqueue/packed.rs#L1016
the behaviour is not correct either way, as there is no check for descriptor specific notification. I guess falling back to always notifying even when VIRTIO_F_EVENT_IDX had been negotiated is the better option, though it may be best to stop advertising support for it.

Reply:
So what you are saying is that the latter two uses should also check for `is_notif_specific`? Or what kind of descriptor-specific notification check is missing?

Reply:
Sorry, I missed that the offset condition is checked inside `is_notif_specific`.

I think ideally, we would check specifically for the LSB (the current behaviour) in `is_notif` (maybe we can change its name to make it more clear, something like `is_generic_notif_enabled`), and change the if statement to be like the one in the first example. This way, the driver would not send a notification if the device asked for notification for a specific descriptor.

However, I think the way `is_notif_specific` is currently implemented, this would not work for batch transfers. If we push a batch past the descriptor the device specified, we will fail to send a notification since the specific notification condition check is at the end and is a equality check, not a greater-equals one.

For this reason, I think we can continue with the new implementation of `is_notif` in the PR, since sending a notification for each dispatch rather than a specific one is a "SHOULD NOT" and not a "MUST NOT", and will guarantee that we do not skip sending a notification when we need to. However, since this means we effectively do not support descriptor specific notifications, it may be preferable not to advertise support for `VIRTIO_F_EVENT_IDX`, too. We can also remove the second part of the "or" expression in the first example, since the new implementation means that if the first operand is false, the second operand is guaranteed to be false, too.

A future PR that correctly implements the descriptor specific notification also for batch dispatches can reenable support for `VIRTIO_F_EVENT_IDX` and implement `is_generic_notif_enabled`, along with other necessary changes.
  
For reference, the relevant section of the specification (VIRTIO Spec. v1.2 § 2.7.10.1 (the section number is probably different for version 1.1)):
> if the VIRTIO_F_EVENT_IDX feature bit is negotiated:  • The driver MUST ignore the lower bit of flags.  • After the driver writes a descriptor index into the available ring:  – If the idx field in the available ring (which determined where that descriptor index was placed) was equal to avail_event, the driver MUST send a notification.  – Otherwise the driver SHOULD NOT send a notification.

(Note: the quote is from the split queue section, as the packed queue section does not go into as much detail. I would assume that they should behave similarly in this specific issue.)

### Alternative that may be outside the scope of this PR
We can rename the existing function, add the implementation in the PR as a new function, use the new function in the `dispatch_batch{,_await}` functions and change the if condition inside the non-batch dispatch to be a or statement like the one in the first example above.

Reply:
I should note that I don't have a lot of confidence in my understanding of the notification part of the specification as I previously focused mostly on the handling of the descriptors.

Reply:
I have implemented it the way we talked about.